File: tck/scala-tck/src/main/scala/kalix/tck/model/action/ActionTckModelImpl.scala
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package kalix.tck.model.action

import akka.NotUsed
import akka.stream.Materializer
import akka.stream.scaladsl.{ Sink, Source }
import kalix.scalasdk.action.Action
import kalix.scalasdk.action.ActionCreationContext

class ActionTckModelImpl(ctx: ActionCreationContext) extends AbstractActionTckModelAction {
  private implicit val mat: Materializer = ctx.materializer()

  override def processUnary(request: Request): Action.Effect[Response] =
    response(request.groups)

  override def processStreamedIn(requestSrc: Source[Request, NotUsed]): Action.Effect[Response] =
    effects.asyncEffect(
      requestSrc
        // collect all requests
        .runWith(Sink.seq)
        // and then create a single resource
        .map(reqs => response(reqs.flatMap(_.groups))))

  override def processStreamedOut(request: Request): Source[Action.Effect[Response], NotUsed] =
    // each group can create a response
    Source(request.groups).map(g => response(g :: Nil))

  override def processStreamed(requestSrc: Source[Request, NotUsed]): Source[Action.Effect[Response], NotUsed] =
    requestSrc.mapConcat(_.groups).map(g => response(g :: Nil))

  def response(groups: Seq[ProcessGroup]): Action.Effect[Response] = {
    val allSteps = groups.flatMap(_.steps).map(_.step)
    val failed = allSteps.reverse.collectFirst { case ProcessStep.Step.Fail(f) =>
      effects.error(f.message)
    }
    val sideEffects =
      allSteps.collect { case ProcessStep.Step.Effect(e) =>
        kalix.scalasdk.SideEffect(components.actionTwoImpl.call(OtherRequest(e.id)), e.synchronous)
      }

    val res =
      failed match {
        case Some(e) => e
        case None =>
          allSteps.reverse
            .collectFirst {
              case ProcessStep.Step.Reply(r)   => effects.reply(Response(r.message))
              case ProcessStep.Step.Forward(f) => effects.forward(components.actionTwoImpl.call(OtherRequest(f.id)))
            }
            .getOrElse(effects.reply(Response.defaultInstance))
      }
    res.addSideEffects(sideEffects)
  }
}

File: tck/java-tck/src/main/java/kalix/tck/model/action/ActionTckModelImpl.java
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package kalix.tck.model.action;

import akka.NotUsed;
import akka.stream.javadsl.Sink;
import akka.stream.javadsl.Source;
import kalix.javasdk.DeferredCall;
import kalix.javasdk.SideEffect;
import kalix.javasdk.action.*;
import kalix.tck.model.action.Action.*;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.CompletionStage;

public class ActionTckModelImpl extends AbstractActionTckModelAction {

  public ActionTckModelImpl(ActionCreationContext creationContext) {}

  public Effect<Response> processUnary(Request request) {
    // Multiple request steps should be combined to give just one response, with subsequent steps
    // taking precedence.
    return response(request.getGroupsList());
  }

  public Effect<Response> processStreamedIn(Source<Request, NotUsed> requests) {
    // All request steps should be combined to produce a single response after the request stream
    // completes.
    CompletionStage<ArrayList<ProcessGroup>> processGroups =
        requests
            .fold(
                new ArrayList<ProcessGroup>(),
                (acc, request) -> {
                  acc.addAll(request.getGroupsList());
                  return acc;
                })
            .runWith(Sink.head(), actionContext().materializer());
    CompletionStage<Effect<Response>> effect = processGroups.thenApply(groups -> response(groups));
    return effects().asyncEffect(effect);
  }

  public Source<Effect<Response>, NotUsed> processStreamedOut(Request request) {
    // The single request may contain multiple grouped steps, each group corresponding to an
    // expected response.
    return Source.from(request.getGroupsList())
        .map(groupList -> response(Collections.singletonList(groupList)));
  }

  public Source<Effect<Response>, NotUsed> processStreamed(Source<Request, NotUsed> requests) {
    // Each request may contain multiple grouped steps, each group corresponding to an expected
    // response.
    return requests
        .mapConcat(request -> request.getGroupsList())
        .map(groupList -> response(Collections.singletonList(groupList)));
  }

  private Effect<Response> response(List<ProcessGroup> groups) {
    Effect<Response> effect = null;
    List<SideEffect> sideEffects = new ArrayList<>();
    // TCK tests expect the logic to be imperative, building something up and then discarding on
    // failure but effect
    // api is not imperative so we have to keep track of failure instead
    boolean didFail = false;
    for (ProcessGroup group : groups) {
      for (ProcessStep step : group.getStepsList()) {
        switch (step.getStepCase()) {
          case REPLY:
            if (!didFail) {
              effect =
                  effects()
                      .reply(
                          Response.newBuilder().setMessage(step.getReply().getMessage()).build());
            }
            break;
          case FORWARD:
            if (!didFail) {
              effect = effects().forward(serviceTwoRequest(step.getForward().getId()));
            }
            break;
          case EFFECT:
            Action.SideEffect sideEffect = step.getEffect();
            sideEffects.add(
                kalix.javasdk.SideEffect.of(
                    serviceTwoRequest(sideEffect.getId()), sideEffect.getSynchronous()));
            break;
          case FAIL:
            effect = effects().error(step.getFail().getMessage());
            didFail = true;
        }
      }
    }
    if (effect == null) {
      effect = effects().reply(Response.getDefaultInstance());
    }
    return effect.addSideEffects(sideEffects);
  }

  private DeferredCall serviceTwoRequest(String id) {
    return components().actionTwoImpl().call(OtherRequest.newBuilder().setId(id).build());
  }
}
